File: apps/reading/src/api/deepResearchSourceCitationPack.ts
/**
 * Deep research source citation pack (pure).
 *
 * Operator vision: reference arxiv, substack, and other knowledge-dense
 * publications in deep research — highest quality DR product. This pure layer
 * builds a citation pack from operator-supplied citation records and selected
 * source families. Never invents citation metadata or remote fetches.
 *
 * remote_fetched is always false.
 */

import {
  DEFAULT_PUBLICATION_CATALOG,
  selectPublicationSources,
  type PublicationFamily,
  type SourceSelectionPack,
} from "./sourcePublicationRegistry";

export type CitationFamily = PublicationFamily;

export interface CitationRecord {
  citation_id: string;
  family: CitationFamily;
  /** Operator-supplied title (never invented). */
  title: string;
  /** Optional stable external id (e.g. arxiv:2301.00001). */
  external_id?: string;
  /** Optional URL — caller-supplied only. */
  url?: string;
  /** Optional year. */
  year?: number | null;
  /** Optional authors string. */
  authors?: string | null;
}

export interface DeepResearchSourceCitationPackInput {
  session_id: string;
  /** Families to include for this research (arxiv, substack, …). */
  requested_families: CitationFamily[];
  /** Operator-supplied citation rows only. */
  citations: CitationRecord[];
  /** When true, drop citations whose family is not in the selection pack. */
  filter_to_selected_families?: boolean;
}

export interface DeepResearchSourceCitationPack {
  session_id: string;
  selection: SourceSelectionPack;
  citations: CitationRecord[];
  citation_count: number;
  families_present: CitationFamily[];
  /** True when ≥1 citation and ≥1 selected family. */
  pack_ready: boolean;
  /** Always false — pure pack never fetches remote publications. */
  remote_fetched: false;
  notes: string[];
  authority: "deep_research_source_citation_pack_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

const VALID_FAMILIES = new Set<CitationFamily>([
  "arxiv",
  "substack",
  "openalex",
  "web",
  "custom",
]);

/**
 * Build a deep-research citation pack from selection + operator citations.
 * Never invents citation rows; never remote-fetches.
 */
export function buildDeepResearchSourceCitationPack(
  input: DeepResearchSourceCitationPackInput,
): DeepResearchSourceCitationPack {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  const session_id = requireNonEmpty(input.session_id, "session_id");
  if (!Array.isArray(input.requested_families)) {
    throw new Error("requested_families must be an array");
  }
  if (!Array.isArray(input.citations)) {
    throw new Error("citations must be an array");
  }

  const notes: string[] = [
    "remote_fetched=false — citation pack is selection + caller records only",
    "citation rows are operator-supplied only (no invent / no scrape)",
  ];

  const selection = selectPublicationSources({
    requested_families: input.requested_families,
    enabled_only: true,
  });
  notes.push(...selection.notes);
  notes.push(`selected_families=${selection.families.join(",") || "(none)"}`);

  const selectedFamilySet = new Set(selection.families);
  const filter =
    input.filter_to_selected_families === undefined
      ? true
      : input.filter_to_selected_families;
  if (typeof filter !== "boolean") {
    throw new Error("filter_to_selected_families must be boolean when set");
  }

  const seen = new Set<string>();
  const citations: CitationRecord[] = [];
  const families_present = new Set<CitationFamily>();

  for (let i = 0; i < input.citations.length; i++) {
    const c = input.citations[i];
    if (!c || typeof c !== "object") {
      throw new Error(`citations[${i}] must be an object`);
    }
    const citation_id = requireNonEmpty(
      c.citation_id,
      `citations[${i}].citation_id`,
    );
    if (seen.has(citation_id)) {
      throw new Error(`duplicate citation_id: ${citation_id}`);
    }
    seen.add(citation_id);

    const family = c.family;
    if (!VALID_FAMILIES.has(family as CitationFamily)) {
      throw new Error(
        `citations[${i}].family must be arxiv|substack|openalex|web|custom`,
      );
    }
    const fam = family as CitationFamily;
    if (filter && selectedFamilySet.size > 0 && !selectedFamilySet.has(fam)) {
      notes.push(
        `citations[${i}] family=${fam} filtered out (not in selected families)`,
      );
      continue;
    }
    if (filter && selectedFamilySet.size === 0) {
      notes.push(
        `citations[${i}] dropped — no families selected (filter_to_selected_families)`,
      );
      continue;
    }

    const title = requireNonEmpty(c.title, `citations[${i}].title`);
    let external_id: string | undefined;
    if (c.external_id != null) {
      external_id = requireNonEmpty(
        c.external_id,
        `citations[${i}].external_id`,
      );
    }
    let url: string | undefined;
    if (c.url != null) {
      url = requireNonEmpty(c.url, `citations[${i}].url`);
      // Lightweight shape check — not a live fetch.
      if (!/^https?:\/\//i.test(url) && !url.startsWith("arxiv:")) {
        throw new Error(
          `citations[${i}].url must be http(s) URL or arxiv: id when set`,
        );
      }
    }
    let year: number | null | undefined = c.year;
    if (year !== undefined && year !== null) {
      if (
        typeof year !== "number" ||
        !Number.isInteger(year) ||
        year < 1000 ||
        year > 3000
      ) {
        throw new Error(
          `citations[${i}].year must be integer year in [1000,3000] when set`,
        );
      }
    }
    let authors: string | null | undefined = c.authors;
    if (authors !== undefined && authors !== null) {
      if (typeof authors !== "string" || !authors.trim()) {
        throw new Error(
          `citations[${i}].authors must be non-empty string when set`,
        );
      }
      authors = authors.trim();
    }

    citations.push({
      citation_id,
      family: fam,
      title,
      ...(external_id !== undefined ? { external_id } : {}),
      ...(url !== undefined ? { url } : {}),
      ...(year !== undefined ? { year } : {}),
      ...(authors !== undefined ? { authors } : {}),
    });
    families_present.add(fam);
  }

  if (input.citations.length === 0) {
    notes.push("no citations supplied — empty pack (no invent citations)");
  } else {
    notes.push(
      `citations_accepted=${citations.length} of ${input.citations.length} supplied`,
    );
  }

  const pack_ready =
    selection.families.length >= 1 && citations.length >= 1;
  if (!pack_ready) {
    if (selection.families.length < 1) {
      notes.push("pack_ready=false — need ≥1 selected publication family");
    } else {
      notes.push("pack_ready=false — need ≥1 accepted citation");
    }
  } else {
    notes.push(
      `pack_ready=true · citations=${citations.length} · families=${[...families_present].join(",")}`,
    );
  }
  notes.push("remote_fetched=false");

  // Ensure selection still reports fetched=false
  if (selection.fetched !== false) {
    throw new Error("selection.fetched must be false");
  }

  return {
    session_id,
    selection,
    citations,
    citation_count: citations.length,
    families_present: [...families_present],
    pack_ready,
    remote_fetched: false,
    notes,
    authority: "deep_research_source_citation_pack_advisory",
  };
}

export function formatDeepResearchSourceCitationPackSummary(
  p: DeepResearchSourceCitationPack,
): string {
  return (
    `citation pack · n=${p.citation_count} · ready=${p.pack_ready} · ` +
    `remote_fetched=false · families=${p.families_present.join(",") || "none"}`
  );
}

export { DEFAULT_PUBLICATION_CATALOG };

File: substrate/deep_research_source_citation_pack.py
"""Deep research source citation pack (pure, advisory).

Builds a citation pack from operator-supplied citation records and selected
publication families (arxiv, substack, …). remote_fetched always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.source_publication_registry import (
    SourcePublicationRegistryError,
    SourceSelectionPack,
    select_publication_sources,
)

VALID_FAMILIES = frozenset({"arxiv", "substack", "openalex", "web", "custom"})


class DeepResearchSourceCitationPackError(ValueError):
    """Fail-closed validation for deep research citation pack."""


@dataclass(frozen=True)
class CitationRecord:
    citation_id: str
    family: str
    title: str
    external_id: str | None
    url: str | None
    year: int | None
    authors: str | None

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {
            "citation_id": self.citation_id,
            "family": self.family,
            "title": self.title,
        }
        if self.external_id is not None:
            out["external_id"] = self.external_id
        if self.url is not None:
            out["url"] = self.url
        if self.year is not None:
            out["year"] = self.year
        if self.authors is not None:
            out["authors"] = self.authors
        return out


@dataclass(frozen=True)
class DeepResearchSourceCitationPack:
    session_id: str
    selection: SourceSelectionPack
    citations: tuple[CitationRecord, ...]
    citation_count: int
    families_present: tuple[str, ...]
    pack_ready: bool
    remote_fetched: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "selection": self.selection.to_dict(),
            "citations": [c.to_dict() for c in self.citations],
            "citation_count": self.citation_count,
            "families_present": list(self.families_present),
            "pack_ready": self.pack_ready,
            "remote_fetched": False,
            "notes": list(self.notes),
            "authority": "deep_research_source_citation_pack_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise DeepResearchSourceCitationPackError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def build_deep_research_source_citation_pack(
    *,
    session_id: object,
    requested_families: object,
    citations: object,
    filter_to_selected_families: object = True,
) -> DeepResearchSourceCitationPack:
    """Build citation pack. Never invents rows; never remote-fetches."""
    sid = _require_nonempty(session_id, field="session_id")
    if not isinstance(requested_families, list):
        raise DeepResearchSourceCitationPackError(
            "requested_families must be an array"
        )
    if not isinstance(citations, list):
        raise DeepResearchSourceCitationPackError("citations must be an array")
    if not isinstance(filter_to_selected_families, bool):
        raise DeepResearchSourceCitationPackError(
            "filter_to_selected_families must be boolean when set"
        )

    notes: list[str] = [
        "remote_fetched=false — citation pack is selection + caller records only",
        "citation rows are operator-supplied only (no invent / no scrape)",
    ]

    try:
        selection = select_publication_sources(
            requested_families=requested_families,
            enabled_only=True,
        )
    except SourcePublicationRegistryError as e:
        raise DeepResearchSourceCitationPackError(str(e)) from e

    notes.extend(selection.notes)
    notes.append(
        f"selected_families={','.join(selection.families) or '(none)'}"
    )
    selected_family_set = set(selection.families)

    if selection.fetched is not False:
        raise DeepResearchSourceCitationPackError(
            "selection.fetched must be false"
        )

    seen: set[str] = set()
    accepted: list[CitationRecord] = []
    families_present: set[str] = set()

    for i, c in enumerate(citations):
        if not isinstance(c, dict):
            raise DeepResearchSourceCitationPackError(
                f"citations[{i}] must be an object"
            )
        cid = _require_nonempty(
            c.get("citation_id"), field=f"citations[{i}].citation_id"
        )
        if cid in seen:
            raise DeepResearchSourceCitationPackError(
                f"duplicate citation_id: {cid}"
            )
        seen.add(cid)

        family = c.get("family")
        if family not in VALID_FAMILIES:
            raise DeepResearchSourceCitationPackError(
                f"citations[{i}].family must be arxiv|substack|openalex|web|custom"
            )
        if filter_to_selected_families and selected_family_set and family not in selected_family_set:
            notes.append(
                f"citations[{i}] family={family} filtered out (not in selected families)"
            )
            continue
        if filter_to_selected_families and not selected_family_set:
            notes.append(
                f"citations[{i}] dropped — no families selected "
                "(filter_to_selected_families)"
            )
            continue

        title = _require_nonempty(c.get("title"), field=f"citations[{i}].title")
        external_id = c.get("external_id")
        if external_id is not None:
            external_id = _require_nonempty(
                external_id, field=f"citations[{i}].external_id"
            )
        url = c.get("url")
        if url is not None:
            url = _require_nonempty(url, field=f"citations[{i}].url")
            if not (
                url.lower().startswith("http://")
                or url.lower().startswith("https://")
                or url.startswith("arxiv:")
            ):
                raise DeepResearchSourceCitationPackError(
                    f"citations[{i}].url must be http(s) URL or arxiv: id when set"
                )
        year = c.get("year")
        if year is not None:
            if (
                isinstance(year, bool)
                or not isinstance(year, int)
                or year < 1000
                or year > 3000
            ):
                raise DeepResearchSourceCitationPackError(
                    f"citations[{i}].year must be integer year in [1000,3000] when set"
                )
        authors = c.get("authors")
        if authors is not None:
            if not isinstance(authors, str) or not authors.strip():
                raise DeepResearchSourceCitationPackError(
                    f"citations[{i}].authors must be non-empty string when set"
                )
            authors = authors.strip()

        accepted.append(
            CitationRecord(
                citation_id=cid,
                family=family,  # type: ignore[arg-type]
                title=title,
                external_id=external_id if external_id is not None else None,
                url=url if url is not None else None,
                year=year if year is not None else None,
                authors=authors if authors is not None else None,
            )
        )
        families_present.add(family)  # type: ignore[arg-type]

    if len(citations) == 0:
        notes.append("no citations supplied — empty pack (no invent citations)")
    else:
        notes.append(
            f"citations_accepted={len(accepted)} of {len(citations)} supplied"
        )

    pack_ready = len(selection.families) >= 1 and len(accepted) >= 1
    if not pack_ready:
        if len(selection.families) < 1:
            notes.append(
                "pack_ready=false — need ≥1 selected publication family"
            )
        else:
            notes.append("pack_ready=false — need ≥1 accepted citation")
    else:
        notes.append(
            f"pack_ready=true · citations={len(accepted)} · "
            f"families={','.join(sorted(families_present))}"
        )
    notes.append("remote_fetched=false")

    return DeepResearchSourceCitationPack(
        session_id=sid,
        selection=selection,
        citations=tuple(accepted),
        citation_count=len(accepted),
        families_present=tuple(sorted(families_present)),
        pack_ready=pack_ready,
        remote_fetched=False,
        notes=tuple(notes),
        authority="deep_research_source_citation_pack_advisory",
    )


__all__ = [
    "CitationRecord",
    "DeepResearchSourceCitationPack",
    "DeepResearchSourceCitationPackError",
    "build_deep_research_source_citation_pack",
]
